src/devices/Dhtxx/DhtSensor.cs

```diff
@@ -195,28 +251,32 @@ public bool ReadData()
         }
 
         // Convertion for DHT11
-        private double GetTempDht11() => IsLastReadSuccessful ? (double)(_dht11Val[2] + _dht11Val[3] / 10) : double.MaxValue;
+        private double GetTempDht11() => IsLastReadSuccessful ? _readBuff[2] + _readBuff[3] / 10.0 : double.MaxValue;
 
-        private double GetHumidityDht11() => IsLastReadSuccessful ? (double)(_dht11Val[0] + _dht11Val[1] / 10) : double.MaxValue;
+        private double GetHumidityDht11() => IsLastReadSuccessful ? _readBuff[0] + _readBuff[1] / 10.0 : double.MaxValue;
 
-        // convertion for DHT22
+        // convertion for DHT12, DHT21, DHT22
         private double GetTempDht22()
         {
             if (IsLastReadSuccessful)
             {
-                var temp = (((_dht11Val[2] & 0x7F) << 8) | _dht11Val[3]) * 0.1F;
+                var temp = _readBuff[2] + (_readBuff[3] & 0x7F) * 0.1;
                 // if MSB = 1 we have negative temperature
-                return ((_dht11Val[2] & 0x80) == 0 ? temp : -temp);
+                return ((_readBuff[3] & 0x80) == 0 ? temp : -temp);
             }
             else
                 return (double.MaxValue);
         }
 
-        private double GetHumidityDht22() => IsLastReadSuccessful ? (double)((_dht11Val[0] << 8) | _dht11Val[1]) * 0.1F : double.MaxValue;
+        private double GetHumidityDht22() => IsLastReadSuccessful ? _readBuff[0] + _readBuff[1] * 0.1 : double.MaxValue;
 
+        /// <summary>
+        /// Cleanup
+        /// </summary>
         public void Dispose()
         {
-            _controller.ClosePin(_pin);
+            _controller?.Dispose();
+            _sensor?.Dispose();
```

Comment:
they should not be null unless you nullify them :smile:

Reply:
What if Dispose() is called many times?😄

Reply:
If initialized with i2c, gpio is null, and on the contrary....